### PR TITLE
Appearance: Add dark preference

### PR DESCRIPTION
**Pull request by cassidyjames**
Continuation of #171, but using AccountsService. CI is failing because it requires Granite 5.4 which we don't have yet.

### CODE REVIEW

File: src/Views/Appearance.vala [hunk 1]
```diff
@@ -76,6 +92,53 @@ public class Appearance : Gtk.Grid {
...
+                "org.freedesktop.Accounts",
+                user_path,
+                GLib.DBusProxyFlags.GET_INVALIDATED_PROPERTIES
+            );
+        } catch (Error e) {
+            warning ("Unable to get AccountsService proxy, color scheme preference may be incorrect");
+        }
+
+        critical (pantheon_act.prefers_color_scheme.to_string ());
+
```
Comment (cassidyjames) [inline src/Views/Appearance.vala:122]: Only the best debugging tools…

```suggestion
```

File: src/Views/Appearance.vala [hunk 2]
```diff
@@ -76,6 +92,51 @@ public class Appearance : Gtk.Grid {
...
+            user_path = accounts_service.find_user_by_name (GLib.Environment.get_user_name ());
+        } catch (Error e) {
+            critical (e.message);
+        }
+
+        try {
+            pantheon_act = GLib.Bus.get_proxy_sync (
+                GLib.BusType.SYSTEM,
+                "org.freedesktop.Accounts",
+                user_path,
```
Comment (donadigo) [inline src/Views/Appearance.vala:114]: After getting the user path, you should probably check if it's not an empty string before passing it to another DBus method.

File: src/Views/Appearance.vala [hunk 3]
```diff
@@ -76,6 +92,51 @@ public class Appearance : Gtk.Grid {
...
+                GLib.DBusProxyFlags.GET_INVALIDATED_PROPERTIES
+            );
+        } catch (Error e) {
+            warning ("Unable to get AccountsService proxy, color scheme preference may be incorrect");
+        }
+
+        // FIXME: This seems… not ideal. Can't we bind this?
+        switch (pantheon_act.prefers_color_scheme) {
```
Comment (donadigo) [inline src/Views/Appearance.vala:122]: Connecting to this DBus interface can fail which is handled by the catch above. However, this here will still execute. Check for `pantheon_act != null` before this switch.

File: src/Views/Appearance.vala [hunk 4]
```diff
@@ -76,6 +92,51 @@ public class Appearance : Gtk.Grid {
 
         update_text_size_modebutton (interface_settings);
 
+        Pantheon.AccountsService? pantheon_act = null;
+        FDO.Accounts? accounts_service = null;
```
Comment (donadigo) [inline src/Views/Appearance.vala:96]: This variable can be reduced in scope. Remove this line and add `FDO.Accounts?` at the beginning of the line after try.

File: src/Views/Appearance.vala (hunk 2 above)
Comment (cassidyjames) [reply src/Views/Appearance.vala:114]: @donadigo if it's an empty string, what should we do?

**Comment by cassidyjames** [comment]
@donadigo I've added null checks and reduced the scope of accounts_service. Lemme know if there's anything else!

File: src/Views/Appearance.vala [hunk 5]
```diff
@@ -38,6 +38,19 @@ public class Appearance : Gtk.Grid {
         row_spacing = 6;
         margin_start = margin_end = 6;
 
+        var dark_label = new Gtk.Label (_("Prefer dark style:"));
```
Comment (danrabbit) [inline src/Views/Appearance.vala:41]: I guess we probably shouldn't even construct this UI unless the DBus API exists. Or at least we shouldn't pack it into the grid

Comment (cassidyjames) [reply src/Views/Appearance.vala:41]: @danrabbit I was wondering that and went back and forth. It seemed to be a little more confusing when the `attach` methods were missing up above, but I can move them and add a comment for clarity.

File: src/Views/Appearance.vala [hunk 6]
```diff
@@ -76,6 +91,57 @@ public class Appearance : Gtk.Grid {
...
+                    "org.freedesktop.Accounts",
+                    user_path,
+                    GLib.DBusProxyFlags.GET_INVALIDATED_PROPERTIES
+                );
+            } catch (Error e) {
+                warning ("Unable to get AccountsService proxy, color scheme preference may be incorrect");
+            }
+        }
+
+        if (pantheon_act != null) {
```
Comment (danrabbit) [inline src/Views/Appearance.vala:122]: So this tells us that the Pantheon service file exists, but not that specific endpoint. So we still get the UI even if the prefer dark style stuff isn't in

```suggestion
        if (((GLib.DBusProxy) pantheon_act).get_cached_property ("PrefersColorScheme") != null) {
```

**Pull request merged**